# Lab | Pandas

In this lab, we will be working with the customer data from an insurance company, which can be found in the CSV file located at the following link: https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file1.csv

The data includes information such as customer ID, state, gender, education, income, and other variables that can be used to perform various analyses.

Throughout the lab, we will be using the pandas library in Python to manipulate and analyze the data. Pandas is a powerful library that provides various data manipulation and analysis tools, including the ability to load and manipulate data from a variety of sources, including CSV files.

### Data Description

- Customer - Customer ID

- ST - State where customers live

- Gender - Gender of the customer

- Education - Background education of customers 

- Customer Lifetime Value - Customer lifetime value(CLV) is the total revenue the client will derive from their entire relationship with a customer. In other words, is the predicted or calculated value of a customer over their entire duration as a policyholder with the insurance company. It is an estimation of the net profit that the insurance company expects to generate from a customer throughout their relationship with the company. Customer Lifetime Value takes into account factors such as the duration of the customer's policy, premium payments, claim history, renewal likelihood, and potential additional services or products the customer may purchase. It helps insurers assess the long-term profitability and value associated with retaining a particular customer.

- Income - Customers income

- Monthly Premium Auto - Amount of money the customer pays on a monthly basis as a premium for their auto insurance coverage. It represents the recurring cost that the insured person must pay to maintain their insurance policy and receive coverage for potential damages, accidents, or other covered events related to their vehicle.

- Number of Open Complaints - Number of complaints the customer opened

- Policy Type - There are three type of policies in car insurance (Corporate Auto, Personal Auto, and Special Auto)

- Vehicle Class - Type of vehicle classes that customers have Two-Door Car, Four-Door Car SUV, Luxury SUV, Sports Car, and Luxury Car

- Total Claim Amount - the sum of all claims made by the customer. It represents the total monetary value of all approved claims for incidents such as accidents, theft, vandalism, or other covered events.


External Resources: https://towardsdatascience.com/filtering-data-frames-in-pandas-b570b1f834b9

## Challenge 1: Understanding the data

In this challenge, you will use pandas to explore a given dataset. Your task is to gain a deep understanding of the data by analyzing its characteristics, dimensions, and statistical properties.

In [7]:
import numpy as np
import pandas as pd

df = pd.read_csv('https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file1.csv')
print(df.head())

  Customer          ST GENDER             Education Customer Lifetime Value  \
0  RB50392  Washington    NaN                Master                     NaN   
1  QZ44356     Arizona      F              Bachelor              697953.59%   
2  AI49188      Nevada      F              Bachelor             1288743.17%   
3  WW63253  California      M              Bachelor              764586.18%   
4  GA49547  Washington      M  High School or Below              536307.65%   

    Income  Monthly Premium Auto Number of Open Complaints     Policy Type  \
0      0.0                1000.0                    1/0/00   Personal Auto   
1      0.0                  94.0                    1/0/00   Personal Auto   
2  48767.0                 108.0                    1/0/00   Personal Auto   
3      0.0                 106.0                    1/0/00  Corporate Auto   
4  36357.0                  68.0                    1/0/00   Personal Auto   

   Vehicle Class  Total Claim Amount  
0  Four-Door Car 

<br><br>
- Identify the dimensions of the dataset by determining the number of rows and columns it contains.

In [9]:
# Dimensions of the dataset: 
rows, columns = df.shape

print(f"Number of rows: {rows}")
print(f"Number of columns: {columns}")

Number of rows: 4008
Number of columns: 11


<br><br>
- Determine the data types of each column and evaluate whether they are appropriate for the nature of the variable. You should also provide suggestions for fixing any incorrect data types.

In [11]:
# Data types of each column
data_types = df.dtypes
print(data_types)

Customer                      object
ST                            object
GENDER                        object
Education                     object
Customer Lifetime Value       object
Income                       float64
Monthly Premium Auto         float64
Number of Open Complaints     object
Policy Type                   object
Vehicle Class                 object
Total Claim Amount           float64
dtype: object


### Suggestions for fixing incorrect data types:
- __Customer__: This variable can be a string.
- __ST__: This variable can be a string.
- __Gender__: This variable can be a string.
- __Education__: This variable can be a string.
- __Customer Lifetime Value__: By definition this is a monetary value, not a percentage.  The best data type for this column is *float* because it may involve decimal places (e.g., $2500.75).
- __Number of Open Complaints__: This should be an *integer* data type and not a date data type. This is because it counts the number of complains, which is inherently a discrete, non-negative quantity.
- __Policy Type__: This variable can be a string.
- __Vehicle Class__: This variable can be a string.
- I think it's also important to add the units as part of the label for each column.

<br><br>
- Identify the number of unique values for each column and determine which columns appear to be categorical. You should also describe the unique values of each categorical column.

In [13]:
# Number of unique values for each column
unique_values = df.nunique()

print(unique_values)

Customer                     1071
ST                              8
GENDER                          5
Education                       6
Customer Lifetime Value      1027
Income                        774
Monthly Premium Auto          132
Number of Open Complaints       6
Policy Type                     3
Vehicle Class                   6
Total Claim Amount            761
dtype: int64


### Categorial columns:

In [24]:
# ST - State where customers live
# Columns with a small number of unique values relative to the number of rows are likely categorical:

columns = ["ST", "GENDER", "Education", "Number of Open Complaints", "Policy Type", "Vehicle Class"]

for col in columns:
    unique_values = df[col].unique()
    print(f"The unique values for '{col}' are: {unique_values}\n")

The unique values for 'ST' are: ['Washington' 'Arizona' 'Nevada' 'California' 'Oregon' 'Cali' 'AZ' 'WA'
 nan]

The unique values for 'GENDER' are: [nan 'F' 'M' 'Femal' 'Male' 'female']

The unique values for 'Education' are: ['Master' 'Bachelor' 'High School or Below' 'College' 'Bachelors' 'Doctor'
 nan]

The unique values for 'Number of Open Complaints' are: ['1/0/00' '1/2/00' '1/1/00' '1/3/00' '1/5/00' '1/4/00' nan]

The unique values for 'Policy Type' are: ['Personal Auto' 'Corporate Auto' 'Special Auto' nan]

The unique values for 'Vehicle Class' are: ['Four-Door Car' 'Two-Door Car' 'SUV' 'Luxury SUV' 'Sports Car'
 'Luxury Car' nan]



<br><br>
- Provide the range of values for numerical columns, and give your insights.

In [27]:
# Only numerical columns

numeric_columns = df.select_dtypes(include = [float,int]).columns.tolist()
print(f"Numerical Columns:\n {numeric_columns}\n")

Numerical Columns:
 ['Income', 'Monthly Premium Auto', 'Total Claim Amount']



__"Customer Lifetime Value"__ and __"Number of Open Complaints"__ are also numerical columns:

In [29]:
numeric_columns.append('Customer Lifetime Value')
numeric_columns.append('Number of Open Complaints')
print(f"Numerical Columns:\n {numeric_columns}\n")

Numerical Columns:
 ['Income', 'Monthly Premium Auto', 'Total Claim Amount', 'Customer Lifetime Value', 'Number of Open Complaints']



In [33]:
# range of values for numerical columns

numeric_columns = df.select_dtypes(include = [float,int]).columns.tolist()
print("\nRange of values for Numerical Columns:")
for col in numeric_columns:
    print(f"\nColumn: {col}")
    print(f"Min: {df[col].min()}, Max: {df[col].max()}")


Range of values for Numerical Columns:

Column: Income
Min: 0.0, Max: 99960.0

Column: Monthly Premium Auto
Min: 61.0, Max: 35354.0

Column: Total Claim Amount
Min: 0.382107, Max: 2893.239678


<br><br>
- Compute summary statistics such as mean, median, mode, standard deviation, and quartiles to understand the central tendency and distribution of the data for numerical columns. You should also provide your conclusions based on these summary statistics.

In [341]:
df.describe()

Income  Monthly Premium Auto  Total Claim Amount
count   1071.000000           1071.000000         1071.000000
mean   39295.701214            193.234360          404.986909
std    30469.427060           1601.190369          293.027260
min        0.000000             61.000000            0.382107
25%    14072.000000             68.000000          202.157702
50%    36234.000000             83.000000          354.729129
75%    64631.000000            109.500000          532.800000
max    99960.000000          35354.000000         2893.239678

#### Analysis
- The average income is approximately 39,296€
- The standard deviation is about 30,469€, indicating significant variability in income levels
- The minimum income is 0, indicating that some customers have no income recorded.
- 25% have an income below 14,072€
- The median income is $36,234 € (half of the observations have an income below this value.)
- 75% have an income below 64,631€
- The maximum income is 99,960€

Conclusions: 
- The income distribution is quite spread out, with a large standard deviation relative to the mean. The presence of a $0 income suggests that some individuals might be unemployed, underreporting income, or have zero recorded income for other reasons.
- The distribution of the 'monthly premium auto' is highly skewed, likely due to some outliers with very high premiums. The median value (83€) is significantly lower than the mean (193€).
- The distribution of total claim amounts appears fairly concentrated around the mean, with a few larger claims pushing the maximum up to nearly 2,900. The minimum claim amount being very low suggests that there are some very small claims in the dataset.

<br><br>
- Compute summary statistics for categorical columns and providing your conclusions based on these statistics.

In [51]:
# Count the Occurrences and Percentage of Each Category

columns = ["ST", "GENDER", "Education", "Number of Open Complaints", "Policy Type", "Vehicle Class"]

for col in columns:
    count = df[col].value_counts()
    percentage = df[col].value_counts(normalize=True) * 100
    result = pd.DataFrame({'Count': count, 'Percentage': percentage})
    print(f"Summary for '{col}':\n{result}\n")

Summary for 'ST':
            Count  Percentage
ST                           
Oregon        320   29.878618
California    211   19.701214
Arizona       186   17.366947
Cali          120   11.204482
Nevada         98    9.150327
Washington     81    7.563025
WA             30    2.801120
AZ             25    2.334267

Summary for 'GENDER':
        Count  Percentage
GENDER                   
F         457   47.903564
M         413   43.291405
Male       39    4.088050
female     28    2.935010
Femal      17    1.781971

Summary for 'Education':
                      Count  Percentage
Education                              
Bachelor                324   30.252101
College                 313   29.225023
High School or Below    296   27.637722
Master                   94    8.776844
Doctor                   37    3.454715
Bachelors                 7    0.653595

Summary for 'Number of Open Complaints':
                           Count  Percentage
Number of Open Complaints                   

### Analysis
- Oregon is the most common state with 29.88% of the customers, followed by California (19.70%) and Arizona (17.37%).
- Females (F) make up 47.90% of the customers, while Males (M) account for 43.29%.
- The data suggests that customers are fairly well-educated, with over 87% having education beyond high school.
- Only a small percentage of customers have more than one complaint (about 12% with 1 complaint and less than 11% with more than 1), indicating relatively high customer satisfaction
- The distribution suggests that the majority of customers own standard vehicles, with sports and luxury cars being much less common.

In [74]:
# Calculation of the most frecuent value (top) for each category

columns = ["ST", "GENDER", "Education", "Number of Open Complaints", "Policy Type", "Vehicle Class"]

for col in columns:
    categorical_mode = df[col].describe().top
    print(f"The top value for the {col} is {categorical_mode}:\n")

The top value for the ST is Oregon:

The top value for the GENDER is F:

The top value for the Education is Bachelor:

The top value for the Number of Open Complaints is 1/0/00:

The top value for the Policy Type is Personal Auto:

The top value for the Vehicle Class is Four-Door Car:



### Exercise 1

The marketing team wants to know the top 5 less common customer locations. Create a pandas Series object that contains the customer locations and their frequencies, and then retrieve the top 5 less common locations in ascending order.

In [86]:
# Count the occurrences of each state
location_counts = df['ST'].value_counts()

# Sort the location_counts in ascending order to get the least common locations first
sorted_locations = location_counts.sort_values(ascending=True)

# top 5 least common locations
top_5_least_common_locations = sorted_locations.head(5)
display(top_5_least_common_locations)

ST
AZ             25
WA             30
Washington     81
Nevada         98
Cali          120
Name: count, dtype: int64

### Exercise 2

The sales team wants to know the total number of policies sold for each type of policy. Create a pandas Series object that contains the policy types and their total number of policies sold, and then retrieve the policy type with the highest number of policies sold.

*Hint:*
- *Using value_counts() method simplifies this analysis.*
- *Futhermore, there is a method that returns the index of the maximum value in a column or row.*


In [28]:
# Your code here
policy_counts = df['Policy Type'].value_counts()
display(policy_counts)

Policy Type
Personal Auto     780
Corporate Auto    234
Special Auto       57
Name: count, dtype: int64

In [34]:
most_common_policy = policy_counts.idxmax()

print(f"Policy type with the highest number of policies sold: {most_common_policy}")

Policy type with the highest number of policies sold: Personal Auto


### Exercise 3

The sales team wants to know if customers with Personal Auto have a lower income than those with Corporate Auto. How does the average income compare between the two policy types?

- Use *loc* to create two dataframes: one containing only Personal Auto policies and one containing only Corporate Auto policies.
- Calculate the average income for each policy.
- Print the results.

In [49]:
# Create two data frames:
personal_auto_df = df.loc[df['Policy Type'] == "Personal Auto"]
corporate_auto_df = df.loc[df['Policy Type'] == "Corporate Auto"]

# average income for each policy type:
average_income_personal_auto = personal_auto_df['Income'].mean()
average_income_corporate_auto = corporate_auto_df['Income'].mean()


print(f"Average income for Personal Auto policies: {round(average_income_personal_auto)}")
print(f"Average income for Corporate Auto policies: {round(average_income_corporate_auto)}")

Average income for Personal Auto policies: 38181
Average income for Corporate Auto policies: 41390


### Bonus: Exercise 4


Your goal is to identify customers with a high policy claim amount.

Instructions:

- Review again the statistics for total claim amount to gain an understanding of the data.
- To identify potential areas for improving customer retention and profitability, we want to focus on customers with a high policy claim amount. Consider customers with a high policy claim amount to be those in the top 25% of the total claim amount. Create a pandas DataFrame object that contains information about customers with a policy claim amount greater than the 75th percentile.
- Use DataFrame methods to calculate summary statistics about the high policy claim amount data. 

*Note: When analyzing data, we often want to focus on certain groups of values to gain insights. Percentiles are a useful tool to help us define these groups. A percentile is a measure that tells us what percentage of values in a dataset are below a certain value. For example, the 75th percentile represents the value below which 75% of the data falls. Similarly, the 25th percentile represents the value below which 25% of the data falls. When we talk about the top 25%, we are referring to the values that fall above the 75th percentile, which represent the top quarter of the data. On the other hand, when we talk about the bottom 25%, we are referring to the values that fall below the 25th percentile, which represent the bottom quarter of the data. By focusing on these groups, we can identify patterns and trends that may be useful for making decisions and taking action.*

*Hint: look for a method that gives you the percentile or quantile 0.75 and 0.25 for a Pandas Series.*

*Hint 2: check `Boolean selection according to the values of a single column` in https://towardsdatascience.com/filtering-data-frames-in-pandas-b570b1f834b9*

In [ ]:
# Your code here